internal/command/launch/state.go

```diff
@@ -157,3 +163,76 @@ func (state *launchState) PlanSummary(ctx context.Context) (string, error) {
 	}
 	return ret, nil
 }
+
+func (state *launchState) validateExtensions(ctx context.Context) error {
+
+	// This is written a little awkwardly with the expectation
+	// that we'll probably need more validation in the future.
+	// When that happens we can just errors.Join(a(), b(), c()...)
+
+	io := iostreams.FromContext(ctx)
+	noConfirm := !io.IsInteractive() || flag.GetBool(ctx, "now")
+
+	org, err := state.Org(ctx)
+	if err != nil {
+		return err
+	}
+
+	validateSupabase := func() error {
```

Comment:
Could it be a top-level function, so that you can write some tests?

Reply:
not sure how easy/viable to test this would be!

essentially the only logic here that isn't prompting to continue is
```go
		response, err := gql.ListAddOns(ctx, client, "supabase")
		if err != nil {
			return fmt.Errorf("failed to list Supabase databases: %w", err)
		}

		// TODO: We'd like to be able to query the user's plan to see if they're on a paid plan.
		//       For now, we'll just nag when they create their second database, every time.

		if len(response.AddOns.Nodes) != 1 {
			// If we're at zero databases, we're within the free plan.
			// If we're at >=2 databases, we know we're on a paid plan.
			// It's only 1 existing database where we need to validate the plan.
			return nil
		}
```
and the test would have to make a synthetic `ListAddOnsResponse` anyway, so the only thing being tested would be `if len(response.AddOns.Nodes) != 1`.

Not sure if that's worth restructuring the code (at this point)
Definitely worth keeping in mind though if we get a richer API that can tell us the plan they're on!